File: infrastructure/db-core/src/main/java/com/nmh/commerce/coupon/CouponRepositoryImpl.java
package com.nmh.commerce.coupon;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Repository;

import java.util.Optional;

@Repository
@RequiredArgsConstructor
public class CouponRepositoryImpl implements CouponRepository {

    private final CouponJpaRepository couponJpaRepository;

    @Override
    public Optional<AbstractCoupon> findById(Long id) {
        return couponJpaRepository.findById(id)
            .map(CouponEntity::toDomain);
    }
}

File: infrastructure/db-core/src/main/kotlin/com/nmh/commerce/coupon/CouponRepositoryImpl.kt
package com.nmh.commerce.coupon

import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Repository

@Repository
class CouponRepositoryImpl(
    private val couponJpaRepository: CouponJpaRepository,
) : CouponRepository {
    override fun findById(id: Long): Coupon =
        couponJpaRepository.findByIdOrNull(id)?.toDomain()
            ?: throw NoSuchElementException("Coupon not found")
}
